GH-35344: [C++][Format] Implementation of the LIST_VIEW and LARGE_LIST_VIEW array formats

felipecrv opened the pull request:
### Rationale for this change

Mailing list discussion: https://lists.apache.org/thread/r28rw5n39jwtvn08oljl09d4q2c1ysvb

### What changes are included in this PR?

Initial implementation of the new format in C++.

### Are these changes tested?

Unit tests being written on every commit adding new functionality. More needs to be implemented for Integration Tests (required) to be implementable.

### Are there any user-facing changes?

A new array format. It should have no impact for users that don't use it.
* Closes: #35344

felipecrv commented:
@bkietz

felipecrv commented:
Issue listing things this initial list-view implementation doesn't cover right away: https://github.com/apache/arrow/issues/37249

bkietz commented on cpp/src/arrow/util/list_util_test.cc:
```diff
@@ -0,0 +1,228 @@
+    std::shared_ptr<ArrayData> result;
+
+    // Empty list-like array
+    ASSERT_OK(builder_->FinishInternal(&result));
```
Why use a builder here instead of ArrayFromJSON?

felipecrv replied:
`ArrayFromJSON` can't create list-views with overlapping ranges and other weird variations that are important for `RangeOfValuesUsed`.

felipecrv replied:
I also think (I'm not sure) I added these before I was able to make the JSON conversions work.

bkietz replied:
Great, could you add comments explaining why these arrays are strange cases (and therefore require more manual construction)

bkietz commented on cpp/src/arrow/array/builder_base.cc:
```diff
@@ -319,10 +324,30 @@ struct DerefConstIterator {
 
 Status ArrayBuilder::AppendScalar(const Scalar& scalar, int64_t n_repeats) {
-  if (!scalar.type->Equals(type())) {
+  if (!AreScalarTypesEquivalent(*scalar.type, *type())) {
```
Where is this used? I'm not sure I like the idea of introducing implicit conversions. For one thing, what if the builder is `list(field("item", int8(), /*nullable=*/false))` but the appended scalar is `list(int8())`? The implicit conversion would drop the guarantee of no nulls from the built array

felipecrv replied:
This simplified work I was doing in compute kernels tremendously. I think that if since we have different physical representation for the same logical types we should start thinking about these conversions. Perhaps not in the ad-hoc way I've done here but with a more robust and general type-checking algorithm.

bkietz replied:
I'm not opposed to the conversion, just to making it a property of the builder. I think if things do break then it'd be possible to recover functionality by extracting the conversion to `template <typename S> Result<S> EquivalentCast(const Scalar&);` (or maybe being more formal and adding `Scalar::View`) and calling that before AppendScalar

felipecrv replied:
I added a commit that reverts these changes. All tests passes. I might float this type equivalence idea in the future when I can build a case for it.

bkietz reviewed (changes requested):
Thanks for working on this, it's looking great!

felipecrv commented on cpp/src/arrow/array/array_nested.cc:
```diff
@@ -137,6 +137,78 @@ Result<std::shared_ptr<typename TypeTraits<TYPE>::ArrayType>> ListArrayFromArray
+  const auto& typed_sizes = checked_cast<const OffsetArrayType&>(sizes);
+
+  auto derived_validity_buffer = std::move(null_bitmap);
+  int64_t null_count_ = null_count;
+  if (offsets.null_count() > 0) {
+    derived_validity_buffer = offsets.null_bitmap();
+    null_count_ = offsets.null_count();
+    // We allow construction from an offsets array containing one extra value.
+    // If that is the case, we might need to discount one null from out_null_count.
+    if (offsets.length() - 1 == sizes.length() && !offsets.IsValid(sizes.length())) {
+      null_count_ -= 1;
```
An initial version of this code compared against `null_count` in the end. I will remove the extra variable now.

felipecrv commented on cpp/src/arrow/array/array_nested.cc:
```diff
@@ -189,11 +261,109 @@ Result<std::shared_ptr<Array>> FlattenListArray(const ListArrayT& list_array,
+  //
+  //   1) does not contain nulls
+  //   2) has sorted offsets
+  //   3) every view is disjoint
```
True. They have to be densely packed. I will change the comment.

felipecrv commented on cpp/src/arrow/array/array_nested.cc:
```diff
@@ -189,11 +261,109 @@ Result<std::shared_ptr<Array>> FlattenListArray(const ListArrayT& list_array,
+    break;
+  }
+  i += 1;
+  for (; i < list_view_array_length; i++) {
```
I tried to write this with a single loop, but separating like this allows the second loop to build up on the invariant that `last_offset` is set which makes it much cleaner and easier to understand.

felipecrv replied:
The first version looks like what I initially written, but it contains one extra boolean test

`if (first_i == kUninitialized) {`

This is exactly what I wanted to get rid of with the two loops. When I looked past the aversion to having multiple loops, I could read the code and informally prove its correctness by induction: first loop establishes the base case and the second loop builds on that. Final check at the end for cases with only nulls or empty list-views.

Can incorporate some of your style and do this?

```cpp
  constexpr int64_t kUninitialized = -1;
  int64_t first_i = kUninitialized;
  offset_type end_offset;
  int64_t i = 0;
  for (; i < list_view_array_length; i++) {
    if (is_null_or_empty(i)) continue;

    first_i = i;
    end_offset = offsets[i] + sizes[i];
    break;
  }
  i += 1;
  for (; i < list_view_array_length; i++) {
    if (is_null_or_empty(i)) continue;

    if (offsets[i] == end_offset) {
      end_offset += sizes[i];
      continue;
    }
    non_null_fragments.push_back(
        SliceArrayWithOffsets(*value_array, offsets[first_i], end_offset));
    first_i = i;
    end_offset = offsets[i] + sizes[i];
  }
  if (first_i != kUninitialized) {
    non_null_fragments.push_back(
        SliceArrayWithOffsets(*value_array, offsets[first_i], end_offset));
  }

```

felipecrv replied:
I can extract a few more functions. Just a minute.

felipecrv commented on cpp/src/arrow/array/validate.cc:
```diff
@@ -699,55 +713,188 @@ struct ValidateArrayImpl {
+    return Status::OK();
+  }
+
+  enum ListViewValidationError {
+    kOk = 0,
+    kOutOfBoundsOffset = 1,
+    kOutOfBoundsSize = 2,
+  };
```
Oh yeah. I can do this instead.

felipecrv replied:
Done. Pushing soon.

felipecrv commented on cpp/src/arrow/array/validate.cc:
```diff
@@ -699,55 +713,188 @@ struct ValidateArrayImpl {
+              if (offset < 0 || offset > offset_limit) {
+                return {kOutOfBoundsOffset, slot};
+              }
+              if (offset + size > offset_limit) {
```
Good catch. I can rewrite this to `if (size > offset_limit - offset)` that has no under/overflow issues because `offset <= offset_limit` after the conditional above has been judged `false`.

felipecrv commented on cpp/src/arrow/builder.cc:
```diff
@@ -219,6 +219,20 @@ struct MakeBuilderImpl {
+  Status Visit(const ListViewType& list_view_type) {
+    std::shared_ptr<DataType> value_type = list_view_type.value_type();
+    ARROW_ASSIGN_OR_RAISE(auto value_builder, ChildBuilder(value_type));
+    out.reset(new ListViewBuilder(pool, std::move(value_builder), type));
```
Done in a fixup commit.

felipecrv commented on cpp/src/arrow/array/array_nested.cc:
```diff
@@ -189,11 +261,109 @@ Result<std::shared_ptr<Array>> FlattenListArray(const ListArrayT& list_array,
+        SliceArrayWithOffsets(*value_array, offsets[first_i], end_offset));
+  }
+
+  // Final attempt to avoid invoking Concatenate().
+  if (non_null_fragments.size() == 1) {
+    return non_null_fragments[0];
+  } else if (non_null_fragments.size() == 0) {
+    return MakeEmptyArray(value_array->type(), memory_pool);
+  }
+
+  return Concatenate(non_null_fragments, memory_pool);
```
Done.

bkietz commented on cpp/src/arrow/array/validate.cc:
```diff
@@ -699,55 +713,179 @@ struct ValidateArrayImpl {
+    kOk = 0,
+    kOutOfBoundsOffset = 1,
+    kOutOfBoundsSize = 2,
+  };
```
```suggestion
```

bkietz commented on cpp/src/arrow/array/concatenate.cc:
```diff
@@ -602,8 +691,11 @@ class ConcatenateImpl {
 Result<std::shared_ptr<Array>> Concatenate(const ArrayVector& arrays, MemoryPool* pool) {
-  if (arrays.size() == 0) {
-    return Status::Invalid("Must pass at least one array");
+  switch (arrays.size()) {
```
:ok_hand:

bkietz commented on cpp/src/arrow/array/concatenate.cc:
```diff
@@ -248,6 +305,38 @@ class ConcatenateImpl {
+    value_ranges.reserve(in_.size());
+    for (const auto& input : in_) {
+      ArraySpan input_span(*input);
+      Range range;
+      ARROW_ASSIGN_OR_RAISE(std::tie(range.offset, range.length),
+                            list_util::internal::RangeOfValuesUsed(input_span));
+      value_ranges.push_back(range);
```
(Not actually a suggestion for this PR, just speculation about future helpers) what would you think of adding a macro for `ASSIGN_OR_RAISE` with structured binding?
```suggestion
      ARROW_ASSIGN_OR_RAISE_LIST((offset, length),
                                 list_util::internal::RangeOfValuesUsed(input_span));
      value_ranges.push_back({offset, length});
```

felipecrv replied:
I don't mind the explicitness of `std::tie(range.offset, range.length)` as pairs (which I use a lot) have the problem that you need to understand the context to be sure about the meaning of the `.second` field — is it a length or the inclusive offset bound.

bkietz reviewed (commented):
I think this is basically ready, just a few more comments:

felipecrv commented:
@bkietz I removed the unused `enum`, squashed the fixups, rebased, and force-pushed to get the latest commits from `main` and fix the conflict (rebase didn't even complain about it).

pitrou commented on cpp/src/arrow/array/builder_nested.h:
```diff
@@ -80,100 +89,91 @@ class BaseListBuilder : public ArrayBuilder {
+  /// \pre if is_valid is false, list_length MUST be 0
+  /// \param is_valid Whether the new list slot is valid
+  /// \param list_length The number of elements in the list
+  Status Append(bool is_valid, int64_t list_length) {
```
Is it really necessary to have the caller pass the list length, or could it simply inferred in the next `Append` or `Finish` call?

felipecrv replied:
I want the list-view builder to allow users to control the sizes they push into the sizes buffer.

Example: I can append logical values `[4, 5, 6]`, `[4, 5]`, and `[4]` by

```
Append(true, 3)
Append(true, 2)
Append(true, 1)
child->Append([4, 5, 6])
```

List of suffixes is also possible with careful alternations of child mutations and size appends.

felipecrv replied:
Not that this is the new parent class of both list and list-view builders. The classic list builders remain with their API `Append(bool)` is still supported.

felipecrv replied:
The builders are mirroring the same sub-typing chain I created for the Arrays.

![image](https://github.com/apache/arrow/assets/207795/0603f862-550a-484d-8940-082122ff5181)

mapleFU replied:
This would be useful, can we add this in comment? Since the interface would be a bit tricky.

pitrou commented on cpp/src/arrow/array/builder_nested.h:
```diff
@@ -40,37 +40,46 @@ namespace arrow {
 
 template <typename TYPE>
-class BaseListBuilder : public ArrayBuilder {
+class ARROW_EXPORT VarLengthListLikeBuilder : public ArrayBuilder {
```
I would keep the `Base` prefix to stress that it's not the final implementation class, e.g. `BaseVarLengthListBuilder`.

felipecrv replied:
```
Base+VarLengthListLikeArray

or 

Base+VarLengthListArray
```

?

pitrou replied:
Fair enough, let's not bikeshed this :-)

pitrou commented on cpp/src/arrow/array/array_nested.h:
```diff
@@ -216,6 +231,170 @@ class ARROW_EXPORT LargeListArray : public BaseListArray<LargeListType> {
+  /// This function does the bare minimum of validation of the offsets/sizes and
+  /// input types. TODO: describe the minimum validation
+  ///
+  /// TODO: review this
```
Is the TODO for you to act on? Or does it require others' opinions?

felipecrv replied:
For me to review. I cleaned it up now.

pitrou commented on cpp/src/arrow/array/array_nested.h:
```diff
@@ -43,30 +43,31 @@ namespace arrow {
 
 template <typename TYPE>
-class BaseListArray;
+class VarLengthListLikeArray;
```
Keep a `Base` prefix here as well?

felipecrv replied:
Waiting for an answer of my question on the thread above.

pitrou commented on cpp/src/arrow/array/validate.cc:
```diff
@@ -699,55 +713,173 @@ struct ValidateArrayImpl {
-      // (ARROW-544)
-      if (data.length > 0) {
-        return Status::Invalid("Non-empty array but offsets are null");
+      // For length 0, an empty offsets buffer is accepted (ARROW-544).
```
I don't think we want to extend this special-case to list views. This is (should be?) historical.
Same for sizes below.

felipecrv replied:
I think it's natural that this issue will affect list views the same way it has affected lists.

felipecrv replied:
But I can start being more strict and wait for it.

felipecrv replied:
Note that `Buffer::Wrap(zero_len_vector)` can lead to a `Buffer` containing `nullptr` or an allocated value depending on whether the `std::vector` has increased in capacity or not (e.g. with a `.reserve()` call).

pitrou commented on cpp/src/arrow/array/validate.cc:
```diff
@@ -699,55 +713,173 @@ struct ValidateArrayImpl {
+    const auto* sizes = data.GetValues<offset_type>(2);
+
+    int64_t slot = 0;
+    if (validity) {
```
You can use `VisitSetBitRuns` which should make this massively simpler. For example:
```c++
VisitSetBitRuns(validity, data.offset, data.length, [&](int64_t run_start, int64_t run_length) {
  for (int i = run_start; i < run_length; ++i) {
    const auto size = sizes[i];
    if (size > 0) {
      const auto offset = offsets[i];
      if (offset < 0 || offset > offset_limit) {
        return OutOfBoundsListViewOffset<offset_type>(i, offset_limit);
      }
      if (size > offset_limit - offset) {
        return OutOfBoundsListViewSize<offset_type>(i, offset_limit);
      }
    } else if (size < 0) {
      return OutOfBoundsListViewSize<offset_type>(i, offset_limit);
    }
  }
});
```

felipecrv replied:
Done in a fixup now. I don't know I didn't use the utility right away.

pitrou commented on cpp/src/arrow/util/list_util.cc:
```diff
@@ -0,0 +1,285 @@
+      break;
+  }
+  DCHECK(!is_var_length_list_like(*input.type));
+  return Status::Invalid("RangeOfValuesUsed: input is not a var-length list-like array");
```
Make this `Status::TypeError` since it's a type error.

felipecrv replied:
Done.

pitrou commented on cpp/src/arrow/util/list_util.cc:
```diff
@@ -0,0 +1,285 @@
+  ARROW_ASSIGN_OR_RAISE(std::shared_ptr<ArrayBuilder> value_builder,
+                        MakeBuilder(value_type, pool));
+  auto list_builder = std::make_shared<ListBuilderType>(pool, value_builder, list_type);
+  RETURN_NOT_OK(list_builder->Reserve(list_view_data->length));
```
Is there a function to get the logical values length of a list view array? If so, we could also call `value_builder->Reserve`.

felipecrv replied:
I created a new utility called `SumOfLogicalListSizes` to use here. Pushing soon.

pitrou commented on cpp/src/arrow/util/list_util.cc:
```diff
@@ -0,0 +1,285 @@
+
+  // To re-use the validity and offsets buffers, a sizes buffer with enough
+  // padding on the beginning is allocated and filled with the sizes after
+  // list_data->offset.
```
You must initialize the initial padding area, otherwise you might leak private data accross IPC or the C Data Interface.

felipecrv replied:
Adding a `memset` here.

pitrou commented on cpp/src/arrow/util/list_util.cc:
```diff
@@ -0,0 +1,285 @@
+  // It's very likely that the first non-null non-empty list-view starts at
+  // offset 0 of the child array.
+  int64_t i = 0;
+  while (i < input.length && (input.IsNull(i) || sizes[input.offset + i] == 0)) {
+    i += 1;
+  }
+  if (i >= input.length) {
+    return 0;
+  }
+  auto min_offset = offsets[input.offset + i];
+  if (ARROW_PREDICT_TRUE(min_offset == 0)) {
+    // Early exit: offset 0 found already.
+    return 0;
+  }
```
This prologue is a bit weird. The early exit when finding offset 0 could also apply inside the loop. 
In other words, you can use `SetBitRunReader` so that you can `break` from inside your own loop.

felipecrv replied:
I will rewrite this, but I only want to compare against 0 in the loop body if `offsets[i] < min_offset`, hence the need to compare `min_offset` against 0 before entering the loop.

felipecrv replied:
I'm going to use an `std::optional`. This is ugly no matter what.

pitrou commented on cpp/src/arrow/util/list_util.cc:
```diff
@@ -0,0 +1,285 @@
+    return validity && !arrow::bit_util::GetBit(validity, i);
+  };
+
+  // It's very likely that the first non-null non-empty list-view starts at
+  // offset zero, so we check that first and potentially early-return a 0.
```
Should update this comment which seems to apply to `MinViewOffset`, not `MaxViewEnd`.

felipecrv replied:
It's gone now.

pitrou commented on cpp/src/arrow/util/list_util_test.cc:
```diff
@@ -0,0 +1,233 @@
+    // These list-views are built manually with the list-view builders instead
+    // of using something like ArrayFromJSON() because we want to test the
+    // RangeOfValuesUsed() function's ability to handle arrays containing
+    // overlapping list-views.
```
Or you can use `ListViewArray::FromArrays` to achieve the same effect :-)

pitrou commented on cpp/src/arrow/array/builder_nested.h:
```diff
@@ -80,100 +89,91 @@ class BaseListBuilder : public ArrayBuilder {
-
   /// \brief Start a new variable-length list slot
   ///
   /// This function should be called before beginning to append elements to the
```
But `list_util_test.cc` showcases appending on the value builder before calling this...

I think the API contract is ok as written, but then our tests shouldn't abuse the API?

felipecrv replied:
This is called to "start a new variable-length list slot". The values added to the values builder before this call are not part of the newly started variable-length list slot.

pitrou commented on cpp/src/arrow/array/builder_nested.h:
```diff
@@ -191,20 +191,129 @@ class BaseListBuilder : public ArrayBuilder {
+
+  Status AppendValues(const offset_type* offsets, const offset_type* sizes,
+                      int64_t length, const uint8_t* valid_bytes) final {
+    // offsets are assumed to be valid, but the first lenght-1 sizes have to be
```
```suggestion
    // offsets are assumed to be valid, but the first length-1 sizes have to be
```

pitrou reviewed (changes requested):
Thanks for this. This looks ok for the most part, but the `VarLengthListLikeBuilder` API seems to favor a bit too much the ease of implementing ListViewBuilder, as opposed to user friendliness and efficiency.

bkietz commented on cpp/src/arrow/array/concatenate.cc:
```diff
@@ -602,8 +691,11 @@ class ConcatenateImpl {
+    case 0:
+      return Status::Invalid("Must pass at least one array");
+    case 1:
+      return arrays[0];
```
https://github.com/apache/arrow/issues/37878

felipecrv replied:
I will undo this. This was done after the initial reviews on this PR. I was also optimizing concatenation of children values when they were a single array, but I will also get rid of that and add an explicit comment about the reason.

felipecrv replied:
I'm keeping the original code in `array_nested.cc` as `FlattenListArray` can totally return non-trimmed slices of arrays.

```cpp
  // Final attempt to avoid invoking Concatenate().
  if (non_null_fragments.size() == 1) {
    return non_null_fragments[0];
  } else if (non_null_fragments.size() == 0) {
    return MakeEmptyArray(value_array->type(), memory_pool);
  }

  return Concatenate(non_null_fragments, memory_pool);
```

So my change was just reverting the commit that added the fast path to `Concatenate()`.

felipecrv commented on cpp/src/arrow/util/list_util.h:
```diff
@@ -0,0 +1,62 @@
+ARROW_EXPORT Result<std::pair<int64_t, int64_t>> RangeOfValuesUsed(
+    const ArraySpan& input);
+
+/// \brief Builds a ListViewArray from a ListArray
```
4 cases fixed and added to a fixup commit.

felipecrv commented on cpp/src/arrow/type.h:
```diff
@@ -990,6 +990,71 @@ class ARROW_EXPORT LargeListType : public BaseListType {
+  static constexpr const char* type_name() { return "list_view"; }
+
+  // ListView can contain any other logical value type
+  explicit ListViewType(const std::shared_ptr<DataType>& value_type)
```
I'm being consistent with `ListType` and `LargeListType` (and all the parametric types). I would like to fix these constructors and type factories systematically instead of diverging from the existing pattern on this PR.

felipecrv replied:
I created an issue for this problem: https://github.com/apache/arrow/issues/37891

felipecrv commented on cpp/src/arrow/testing/random.h:
```diff
@@ -438,6 +438,23 @@ class ARROW_TESTING_EXPORT RandomArrayGenerator {
+  /// \param[in] values The underlying values array
+  /// \param[in] size The size of the generated list array
+  /// \param[in] null_probability the probability of a list value being null
+  /// \param[in] force_empty_nulls if true, null list entries must have 0 length
```
In a fixup now.

felipecrv commented on cpp/src/arrow/scalar_test.cc:
```diff
@@ -394,6 +394,10 @@ class TestRealScalar : public ::testing::Test {
 
+  void TestListViewOf() { TestListOf<ListViewScalar>(list_view(type_)); }
+
+  void TestLargeListViewOf() { TestListOf<LargeListViewScalar>(large_list_view(type_)); }
```
I missed a list of types here, but after adding the list-view types, `MakeNullScalar` gets tested by `CheckMakeNullScalar`.

![image](https://github.com/apache/arrow/assets/207795/7cb5236c-f3d3-490b-b5d6-bcf6578952a3)

felipecrv commented on cpp/src/arrow/ipc/json_simple.cc:
```diff
@@ -555,8 +563,9 @@ class ListConverter final : public ConcreteConverter<ListConverter<TYPE>> {
-    RETURN_NOT_OK(builder_->Append());
     // Extend the child converter with this JSON array
+    ARROW_ASSIGN_OR_RAISE(auto size, this->SizeOfJSONArray(json_obj));
+    RETURN_NOT_OK(builder_->Append(true, size));
```
Do you have a name suggestion for "Add a valid or invalid list of a given size"? Note that, per spec, nulls lists can have non-zero sizes in which case filler values are added to the child array.

felipecrv commented on cpp/src/arrow/integration/json_internal.cc:
```diff
@@ -1492,6 +1506,14 @@ class ArrayReader {
+
+  Status Visit(const LargeListViewType& type) {
+    return Status::NotImplemented("large list-view in JSON");
+  }
```
I implemented it for Go. For C++, should I have it in this PR or a separate PR is fine?

felipecrv replied:
Or should I do it in another PR since this one is quite overwhelming to review?

pitrou replied:
Yes, you can.

felipecrv commented on cpp/src/arrow/compute/kernels/scalar_if_else_benchmark.cc:
```diff
@@ -67,11 +67,15 @@ struct GetBytesProcessedVisitor {
+    total_bytes += num_offsets * sizeof(OffsetType);
+    // NOTE: the sizes buffer is not counted when type is a list-view as that
+    // can make the throughput numbers look better just because the sizes add
+    // increase the number of bytes in the input.
```
I'm fixing a typo here: ~add~ increase

felipecrv commented on cpp/src/arrow/builder.cc:
```diff
@@ -219,6 +219,20 @@ struct MakeBuilderImpl {
+  Status Visit(const LargeListViewType& large_list_view_type) {
+    std::shared_ptr<DataType> value_type = large_list_view_type.value_type();
+    ARROW_ASSIGN_OR_RAISE(auto value_builder, ChildBuilder(value_type));
+    out.reset(new LargeListViewBuilder(pool, std::move(value_builder), type));
```
In a fixup.

felipecrv commented on cpp/src/arrow/array/concatenate_test.cc:
```diff
@@ -203,6 +203,54 @@ TEST_F(ConcatenateTest, LargeListType) {
+    std::vector<int32_t> sizes_vector;
+    sizes_vector.reserve(size);
+    for (int32_t i = 0; i < size; ++i) {
+      sizes_vector.push_back(offsets_vector[i + 1] - offsets_vector[i]);
```
Done locally. Tests passing.

felipecrv commented on cpp/src/arrow/array/concatenate_test.cc:
```diff
@@ -203,6 +203,54 @@ TEST_F(ConcatenateTest, LargeListType) {
+}
+
+TEST_F(ConcatenateTest, LargeListViewType) {
+  Check([this](int32_t size, double null_probability, std::shared_ptr<Array>* out) {
```
New commit added de-duplicating for both lists and list-views.

felipecrv commented on cpp/src/arrow/array/concatenate_test.cc:
```diff
@@ -203,6 +203,54 @@ TEST_F(ConcatenateTest, LargeListType) {
+
+    std::shared_ptr<Array> offsets;
+    auto offsets_vector = this->Offsets<int32_t>(values_size, size);
+    offsets_vector.front() = 0;
+    offsets_vector.back() = values_size;
```
Unnecessary indeed. Removing.

felipecrv commented on cpp/src/arrow/util/list_util.cc:
```diff
@@ -0,0 +1,285 @@
+  }
+
+  // Slow path: scan the buffers entirely.
+  arrow::internal::VisitSetBitRunsVoid(
```
I just rewrote everything now.

felipecrv commented on cpp/src/arrow/array/builder_nested.h:
```diff
@@ -191,20 +191,129 @@ class BaseListBuilder : public ArrayBuilder {
+  /// This function should be called before beginning to append elements to the
+  /// value builder
+  ///
+  /// Prefer Append(is_valid, 0) as that works correctly for list-view types
```
I will delete this paragraph as it doesn't make sense here. But I will clarify why the call to `BASE::Append` works.

felipecrv commented on cpp/src/arrow/array/builder_nested.h:
```diff
@@ -191,20 +191,129 @@ class BaseListBuilder : public ArrayBuilder {
+    // consistent with the offsets to rule out the possibility that the caller
+    // is passing sizes that could work if building a list-view, but don't work
+    // on building a list that requires offsets to be non-decreasing.
+    if (sizes) {
```
> What is the point of exposing this API on ListBuilder if sizes endures costly validation but is then entirely ignored? It doesn't seem to me like a very good idea, as builders are expected to be efficient.

`ListBuilder` still exposes `AppendValues(offsets, length, valid_bytes)` with no loss of efficiency. `AppendValues` with `sizes` is overridden here so that we fulfill the generic contract of building var-length lists (requiring offsets + sizes).

The expensive validation exists to catch people making mistakes of passing invalid `sizes` through the generic interface. They can gain more efficiency by passing `nullptr` on the sizes or using the `ListViewBuilder` directly. 

> And if we want some compatible vector-append, then I suggest to do the reverse:
expose a AppendValues(const offset_type* offsets, int64_t length, const uint8_t* valid_bytes) and let ListViewBuilder compute the sizes.

Doing that has its problems as well: I can't delegate to a version that takes explicit `sizes` in list-view builders without allocating a temporary buffer. But I will implement like that and have you decide which way is better.

felipecrv replied:
As I tried to rewrite this I remembered why I didn't go with the version that infers sizes from offsets (that's what I tried at first): it's not possible to know if the offsets being passed contain the last offset of the array, so the sizes can't really be inferred.

This is why the `sizes` validation doesn't care about the `size[length - 1]`.

pitrou replied:
Hmm, I'm coming back to this discussion and I'm struggling to understand this again.

This seems to hint that List builders and ListView builders are different enough that they should probably be different beasts entirely instead of trying to maximize API and code sharing. What do you think?

felipecrv replied:
When I started this PR, I had everything completely separate and then I saw myself writing a lot of tedious repetitive code. Then I spent quite some time refactoring everything from the beginning to leverage the commonalities between lists and list-views. I'm aware the builder implementation is not so clear, but the complexity of implementation pays itself by the ease of writing general code dealing with lists. This was proven in the unit tests code and will be even more important when we start talking about compute kernels dealing with lists.

felipecrv replied:
> ...then the third size (3) is incorrect yet silently accepted.

Yes. The validations aren't bullet-proof, just best-effort.

> Can we perhaps guard the validation with #ifndef NDEBUG?

Sure. The point of the validation is to catch some mistakes. Debug assert is fine -- `AppendValues` implementations put a lot of trust in the caller as they are.

felipecrv replied:
I'm rewording the comments a bit to promise less about the validation.

felipecrv replied:
I also wrapped the validation so it happens only in debug builds.

felipecrv commented:
@pitrou if you prefer that I rebase the PR and squash these `fixup!` commits, tell me. I'm leaving it like this to give you a chance to review the commits I added after your first review one by one.

bkietz commented on cpp/src/arrow/c/bridge.cc:
```diff
@@ -444,6 +444,10 @@ struct SchemaExporter {
 
   Status Visit(const LargeListType& type) { return SetFormat("+L"); }
 
+  Status Visit(const ListViewType& type) { return SetFormat("+lv"); }
+
+  Status Visit(const LargeListViewType& type) { return SetFormat("+Lv"); }
```
Looks like the vote has established
```suggestion
  Status Visit(const ListViewType& type) { return SetFormat("+vl"); }

  Status Visit(const LargeListViewType& type) { return SetFormat("+vL"); }
```

felipecrv replied:
Fixing up the commit changing the bridge code.

bkietz commented on cpp/src/arrow/ipc/writer.cc:
```diff
@@ -430,6 +492,36 @@ class RecordBatchSerializer {
+      out_->body_buffers.push_back(std::move(value_sizes));
+    }
+
+    --max_recursion_depth_;
```
Nit: please move this to just above the VisitArray call

bkietz commented on cpp/src/arrow/testing/random.cc:
```diff
@@ -594,6 +595,217 @@ std::shared_ptr<Array> OffsetsFromLengthsArray(OffsetArrayType* lengths,
+  const auto& value = metadata->value(index);
+  T output{};
+  if (!internal::ParseValue<ArrowType>(value.data(), value.length(), &output)) {
+    ABORT_NOT_OK(Status::Invalid("Could not parse ", key, " = ", value));
```
Nice helper
```suggestion
    ABORT_NOT_OK(Status::Invalid("Could not parse ", key, " = ", value, " as ", ArrowType::type_name()));
```

bkietz reviewed (changes requested):
Just a few comments, this is looking good

felipecrv commented on cpp/src/arrow/array/concatenate.cc:
```diff
@@ -159,16 +166,67 @@ Status PutOffsets(const std::shared_ptr<Buffer>& src, Offset first_offset, Offse
+                                  const std::vector<Range>& value_ranges,
+                                  MemoryPool* pool, std::shared_ptr<Buffer>* out) {
+  const int64_t out_size = SumBufferSizes(buffers);
+  if (out_size > std::numeric_limits<offset_type>::max()) {
+    return Status::Invalid("offset overflow while concatenating arrays");
+  }
```
Indeed. Fixing it in a dedicated commit now.

I think this file would be less confusing if less math was done on buffer sizes (can't be trusted) and more reliance on the offset and length of arrays and array slices.

felipecrv commented:
Waiting for @pitrou to take a look before I squash the fixups and rebase.

pitrou commented:
Well, it's probably too late for 14.0.0 regardless, since @raulcd has already got the release branch.

raulcd commented:
There seems to be some conflicts with main now :( you'll have to rebase and fix those.

felipecrv commented:
> There seems to be some conflicts with main now :( you'll have to rebase and fix those.

Rebased and squashed the fixups now.

felipecrv commented:
To clarify, I didn't mean merging to the release branch, but to `main`.

mapleFU commented on cpp/src/arrow/type_traits.cc:
```diff
@@ -67,21 +67,23 @@ int RequiredValueAlignmentForBuffer(Type::type type_id, int buffer_index) {
-    case Type::LIST:  // Offsets may be cast to int32_t*, data is in child array
-    case Type::MAP:   // This is a list array
+    case Type::LIST:       // Offsets may be cast to int32_t*, data is in child array
+    case Type::LIST_VIEW:  // Same as LIST
```
Offsets and sizes?

felipecrv replied:
No need to go into so much detail. Take this to mean that "aligment requirements of numeric buffers are the same between list and list-views".

mapleFU replied:
(I've checkout the largelist, and it has, so maybe note here is proper?)

felipecrv replied:
Fair enough. Made them all consistent, but dropped the comment about child values from the `LIST` case.

mapleFU commented on cpp/src/arrow/util/list_util.cc:
```diff
@@ -0,0 +1,349 @@
+      }
+    }
+  }
+  return min_offset.value_or(0);
```
Hmmm in this case(.i.e `min_offset == 0`) what does it means for return 0?

felipecrv replied:
If all the list-views are null or empty, the `min_offset` option will be empty and I return `0` which is safe because the `MaxViewEnd` will also be 0 in this case.

There is an opportunity of optimization here that I hadn't noticed before: I can return the option, and if it's empty I don't have to call `MaxViewEnd`.

felipecrv commented on cpp/src/arrow/array/util.cc:
```diff
@@ -280,6 +279,17 @@ class ArrayDataEndianSwapper {
+    RETURN_NOT_OK(SwapOffsets<int32_t>(1));
+    RETURN_NOT_OK(SwapOffsets<int32_t>(2));
+    return Status::OK();
+  }
+  Status Visit(const LargeListViewType& type) {
```
`clang-format` doesn't put a line break here and I did it like this to match the two functions for `List` and `LargeList` right above it.

mapleFU commented on cpp/src/arrow/array/concatenate.cc:
```diff
@@ -97,10 +99,18 @@ Status ConcatenateBitmaps(const std::vector<Bitmap>& bitmaps, MemoryPool* pool,
   return Status::OK();
 }
 
+int64_t SumBufferSizes(const BufferVector& buffers) {
```
Would `SizeInBytes` better here?

felipecrv commented on cpp/src/arrow/array/array_nested.h:
```diff
@@ -216,6 +231,172 @@ class ARROW_EXPORT LargeListArray : public BaseListArray<LargeListType> {
+  /// that project views into the values array.
+  ///
+  /// This function does the bare minimum of validation of the offsets/sizes and
+  /// input types. TODO: describe the minimum validation
```
I could swear I had removed this.

mapleFU commented:
I'm a bit busy this two days(And actually I'm not so familiar with these part of code TAT) . I'll take a look carefully this weekend :-)

mapleFU commented on cpp/src/arrow/array/array_nested.cc:
```diff
@@ -189,11 +260,113 @@ Result<std::shared_ptr<Array>> FlattenListArray(const ListArrayT& list_array,
+  if (list_view_array.null_count() == 0) {
+    bool sorted_and_disjoint = true;
+    for (int64_t i = 1; sorted_and_disjoint && i < list_view_array_length; ++i) {
+      sorted_and_disjoint &=
+          sizes[i - 1] == 0 || offsets[i] - offsets[i - 1] == sizes[i - 1];
+    }
```
Why do we skip offset check when `sizes[i - 1] == 0`?

```
offsets: 1 2 3 2 3
sizes:    1 1 0 1 1
```

does this introduce overlap and break (3)?

felipecrv replied:
You're right. I'm pushing a commit with a rewrite of this entire function now + more tests.

felipecrv replied:
An advantage of skipping the check is that when `size[i]==0` we don't have to require the offset to be anything specific as long as the next list-view starts right where the previous non-empty list-view ended.

Your comment made me realize one problem though: if the list-view arrays starts or ends with an empty list-view I could potentially use very random offsets to construct the slice at the end. The new implementation reduces the number of special cases, has an informal proof of correctness by induction, and more tests.

mapleFU commented on cpp/src/arrow/array/data.cc:
```diff
@@ -233,9 +233,22 @@ BufferSpan OffsetsForScalar(uint8_t* scratch_space, offset_type value_size) {
+                                                           offset_type value_size) {
+  auto* offsets = scratch_space;
+  auto* sizes = scratch_space + sizeof(offset_type);
+  reinterpret_cast<offset_type*>(offsets)[0] = 0;
+  reinterpret_cast<offset_type*>(sizes)[0] = value_size;
```
could `mutable_data_as()` used here?

felipecrv replied:
Yes. I will do it now. These utilities were added long after I started this PR. 💀

felipecrv replied:
I'm going to replace many `reinterpret_casts` with `mutable_data_as` and `data_as` but this one can't because `scratch_space` is just a pointer, not a `Buffer`.

mapleFU commented on cpp/src/arrow/array/array_nested.cc:
```diff
@@ -189,11 +260,113 @@ Result<std::shared_ptr<Array>> FlattenListArray(const ListArrayT& list_array,
+                                const ArrayData& data) {
+  DCHECK(is_list_view(data.type->id()));
+  std::vector<std::shared_ptr<Buffer>> buffers = {nullptr, data.buffers[2]};
+  auto offsets_data =
```
Rename to `sizes_data`?

mapleFU replied:
Hmm I also say `SwapOffsets` in latter code, maybe offset is also ok here.

felipecrv replied:
This is a good rename. I'm applying it. It's OK to pass sizes to `SwapOffsets` — it should be obvious to the reader that swapping endianness of sizes is the same operation as swapping the endianness of offsets.

mapleFU commented on cpp/src/arrow/array/validate.cc:
```diff
@@ -699,57 +713,147 @@ struct ValidateArrayImpl {
+ private:
+  /// \pre basic validation has already been performed
+  template <typename offset_type>
+  Status FullyValidateOffsets(int64_t offset_limit) {
```
So this is for List rather than ListView?

felipecrv replied:
Yes. The ListView function is `FullyValidateOffsetsAndValues` as they have to be validated together to guarantee `offset[i] + size[i] <= values.length()`.

felipecrv commented on cpp/src/arrow/array/array_nested.h:
```diff
@@ -216,6 +231,171 @@ class ARROW_EXPORT LargeListArray : public BaseListArray<LargeListType> {
+  /// This function does the bare minimum of validation of the offsets/sizes and
+  /// input types.
+  ///
+  /// Offsets of an Array's null bitmap can be present or an explicit
+  /// null_bitmap, but not both.
```
Yeah. Changing it now for both lists and list-views.

felipecrv commented on cpp/src/arrow/array/array_nested.cc:
```diff
@@ -137,6 +137,77 @@ Result<std::shared_ptr<typename TypeTraits<TYPE>::ArrayType>> ListArrayFromArray
+    return Status::TypeError("List sizes must be ", OffsetArrowType::type_name());
+  }
+
+  if (offsets.offset() != sizes.offset()) {
+    return Status::Invalid("List offsets and sizes must have the same offset");
+  }
```
It's now noted in the docstrings.

felipecrv commented:
@mapleFU I don't understand this comment https://github.com/apache/arrow/pull/35345#discussion_r1359476421 can you clarify?

felipecrv commented on cpp/src/arrow/array/array_nested.cc:
```diff
@@ -137,6 +137,77 @@ Result<std::shared_ptr<typename TypeTraits<TYPE>::ArrayType>> ListArrayFromArray
+  if (offsets.offset() != sizes.offset()) {
+    return Status::Invalid("List offsets and sizes must have the same offset");
+  }
+  const int64_t offset = sizes.offset();
```
Fair. I'm renaming it to `array_offset`.

felipecrv commented:
@pitrou PTAL when you can.

mapleFU commented:
Well I think build the List with size is ok. But can we add some comment for how to use that? Since that would be a bit tricky.

felipecrv commented:
> Well I think build the List with size is ok. But can we add some comment for how to use that? Since that would be a bit tricky.

Sure. Added more documentation.

felipecrv commented on cpp/src/arrow/array/builder_nested.h:
```diff
@@ -115,8 +138,16 @@ class ARROW_EXPORT VarLengthListLikeBuilder : public ArrayBuilder {
 
+  /// \brief Append an empty list slot
+  ///
+  /// \post Another call to Append* or Finish should be made before appending to
+  /// the values builder to ensure list slot remains empty
```
Note that this was already an unwritten contract when there were only regular list builders.

mapleFU commented:
I'm a bit busy on these days, I'll take a careful round before this Sunday

mapleFU commented on cpp/src/arrow/array/concatenate.cc:
```diff
@@ -160,16 +166,69 @@ Status PutOffsets(const std::shared_ptr<Buffer>& src, Offset first_offset, Offse
+// The concatenated child array does not contain values from the beginning
+// if they are not referenced to by any view.
+template <typename offset_type>
+Status ConcatenateListViewOffsets(const BufferVector& buffers,
```
A quick stupid question: would the size here 1 greater than `sizes`?

felipecrv replied:
No. Since these are the list-view offsets.

mapleFU reviewed (commented):
Rest LGTM

Disclaimer: I'm not so familiar with this part of code, and I didn't review the unit tests

felipecrv commented on cpp/src/arrow/array/array_nested.h:
```diff
@@ -120,10 +135,14 @@ class ARROW_EXPORT ListArray : public BaseListArray<ListType> {
-  /// null_bitmap, but not both.
+  /// If a null_bitmap is not provided, the nulls will be inferred from the offsets' or
+  /// sizes' null bitmap. Only one of these two is allowed to have a null bitmap. But if a
+  /// null_bitmap is provided, the offsets array and the sizes array can't have nulls.
```
Copy and pasta mistake. Fixing now.

felipecrv commented on cpp/src/arrow/array/concatenate.cc:
```diff
@@ -113,33 +122,30 @@ Status ConcatenateOffsets(const BufferVector& buffers, MemoryPool* pool,
-  ARROW_ASSIGN_OR_RAISE(*out, AllocateBuffer((out_length + 1) * sizeof(Offset), pool));
-  auto dst = reinterpret_cast<Offset*>((*out)->mutable_data());
+  const int64_t out_size_in_bytes = SumBufferSizesInBytes(buffers);
+  ARROW_ASSIGN_OR_RAISE(*out, AllocateBuffer(sizeof(Offset) + out_size_in_bytes, pool));
```
This is used only for lists. Not list-views and lists need one extra offset.

mapleFU reviewed (approved):
LGTM

Disclaimer: I just review the part of impl and didn't review all tests carefully.

pitrou commented on cpp/src/arrow/testing/random.cc:
```diff
@@ -811,6 +1021,12 @@ std::shared_ptr<Array> RandomArrayGenerator::ArrayOf(const Field& field, int64_t
     return *ARRAY_TYPE::FromArrays(field.type(), *offsets, *values);                 \
   }
 
+#define GENERATE_LIST_VIEW_CASE(ARRAY_TYPE)                                           \
```
Can you `#undef` it afterwards?

pitrou commented on cpp/src/arrow/util/list_util.h:
```diff
@@ -0,0 +1,75 @@
+namespace list_util {
+
+/// \brief Get the child array holding the values from a List or ListView array
+inline const ArraySpan& ValuesArray(const ArraySpan& span) { return span.child_data[0]; }
```
This honestly looks like a pointless indirection to me, though I wouldn't strongly oppose it either.

(also, should it be `constexpr`?)

felipecrv replied:
I'm removing it. Unlike the `ree_util::ValuesArray` , I haven't used it.

pitrou commented on cpp/src/arrow/util/list_util.h:
```diff
@@ -0,0 +1,75 @@
+/// \return The sum of all list or list-view sizes
+ARROW_EXPORT Result<int64_t> SumOfLogicalListSizes(const ArraySpan& input);
+
+/// \brief Build a ListViewArray from a ListArray
+ARROW_EXPORT Result<std::shared_ptr<ListViewArray>> ListViewFromList(
+    const ListArray& source, MemoryPool* pool);
+
+/// \brief Build a LargeListViewArray from a LargeListArray
+ARROW_EXPORT Result<std::shared_ptr<LargeListViewArray>> ListViewFromList(
+    const LargeListArray& source, MemoryPool* pool);
+
+/// \brief Build a ListArray from a ListViewArray
+ARROW_EXPORT Result<std::shared_ptr<ListArray>> ListFromListView(
+    const ListViewArray& source, MemoryPool* pool);
+
+/// \brief Build a LargeListArray from a LargeListViewArray
+ARROW_EXPORT Result<std::shared_ptr<LargeListArray>> ListFromListView(
+    const LargeListViewArray& source, MemoryPool* pool);
```
It would seem more logical to me to have these as static methods of `ListArray` and `ListViewArray`.

felipecrv replied:
I tried it initially, but it brings so many dependencies to `array_nested.cc` (e.g. `MakeBuilder`) so I think it's preferable to keep the list+list-view stuff apart from the core array definitions.

Another option is having these here and calls from `array_nested.cc`. Is that OK? Or should I move all this code into `array_nested.cc`?

pitrou replied:
I don't think it's a problem to use a Builder in `array_nested.cc`. It would be a problem to start including builder APIs from `array_nested.h`.

felipecrv replied:
Moved all the conversion functions to the array classes now. And the tests are moved from list_util_test to list_array_test

pitrou commented on cpp/src/arrow/util/list_util.cc:
```diff
@@ -0,0 +1,353 @@
+  const auto* sizes = input.GetValues<offset_type>(2);
+
+  // Make an access to the sizes buffer only when strictly necessary.
+#define MINIMIZE_MIN_VIEW_OFFSET(i)             \
```
We should ideally keep macros trivial. Here, it can be replaced with a capturing lambda.

pitrou commented on cpp/src/arrow/util/list_util.cc:
```diff
@@ -0,0 +1,353 @@
+  }
+
+  std::optional<offset_type> min_offset;
+  if (validity == nullptr) {
```
Can simply use `VisitSetBitRunsVoid` which automates this scaffolding:
```c++
  VisitSetBitRunsVoid(validity, input.offset, input.length(), [&](int64_t start, int64t_ length) { ... });
```

felipecrv replied:
How can I early return if I find a `0`? (the most common case)

pitrou replied:
Ah, sorry, I had missed the early return. Scratch that :-(

pitrou commented on cpp/src/arrow/util/list_util.cc:
```diff
@@ -0,0 +1,353 @@
+  const auto* sizes = input.GetValues<offset_type>(2);
+
+  // Early-exit: 64-bit overflow detected. This is not possible on a valid list-view,
+  // but we return the maximum possible value to avoid undefined behavior.
```
Unless there's a specific condition where this can be called on invalid data (for example when reading IPC data), we don't care that UB would happen if the list-view is invalid.

felipecrv replied:
This is a public API. I can't guarantee on what kind of data it will be called.

felipecrv replied:
Pushing a commit that removes this check. I think I'm finally understanding where we are strict about validations and where we can assume valid as pre-conditions.

pitrou commented on cpp/src/arrow/testing/random.cc:
```diff
@@ -608,6 +609,218 @@ std::shared_ptr<Array> OffsetsFromLengthsArray(OffsetArrayType* lengths,
+}
+
+template <typename ArrayType, typename RAG>
+Result<std::shared_ptr<Array>> ArrayOfListView(RAG& self, const Field& field,
```
Hmm... why do `RandomListView` and `ArrayOfListView` use two different approaches? This is more code to maintain and understand.

(I'm not sure which approach is better, but ideally we would decide on a common one)

felipecrv replied:
It is indeed a problem, but note that regular list generators are also split in 2 versions. I started this PR with an unified version, but the code was too complicated (the parameterization is different for some reason) and after your first review I added the `random.cc: Simplify and split the random generator into two algorithms` commit.

I mentioned this duplication in the issues about `concatenate.cc` also having yet another random generator: https://github.com/apache/arrow/issues/38656#issuecomment-1806507591

pitrou replied:
Could these at least use a similar `coverage` parameter? I don't think `sparsity` brings much to the table, and the fact that its description needs 8 lines of text probably means it won't get much use in practice.

felipecrv replied:
I was planning to use the sparsity concept in benchmarks as the tighter packing of the data could affect performance, but I will remove it for now and stash the code somewhere.

pitrou commented on cpp/src/arrow/testing/random.cc:
```diff
@@ -608,6 +609,218 @@ std::shared_ptr<Array> OffsetsFromLengthsArray(OffsetArrayType* lengths,
+/// \param[in] seed The seed for the random number generator
+/// \param[in,out] data The array to shuffle
+template <typename ListViewType>
+void ShuffleListViewDataInPlace(SeedType seed, ArrayData& data) {
```
Ideally we would avoid mutable references.
```suggestion
void ShuffleListViewDataInPlace(SeedType seed, ArrayData* data) {
```

felipecrv replied:
Yeah, that's the style. Fixing.

pitrou commented on cpp/src/arrow/testing/random.cc:
```diff
@@ -608,6 +609,218 @@ std::shared_ptr<Array> OffsetsFromLengthsArray(OffsetArrayType* lengths,
+  using UniformDist = std::uniform_int_distribution<int64_t>;
+  UniformDist dist;
+  for (int64_t i = data.length - 1; i > 0; --i) {
+    const auto j = dist(rng, UniformDist::param_type(0, i));
```
I didn't know this two-operand form :-o

felipecrv replied:
I got it from cppreference and it seems to be The Right Way :tm: to change the sampling range instead of creating a new distribution every time.

pitrou reviewed (commented):
I didn't take a look at everything yet, sorry, but here are some early comments.

felipecrv commented on cpp/src/arrow/array/array_nested.h:
```diff
@@ -105,6 +103,23 @@ class BaseListArray : public Array {
+
+  const TypeClass* list_type() const { return this->var_length_list_like_type(); }
+
+  offset_type value_length(int64_t i) const final {
```
Yes, unlocking inlining of the call. And all the other complicated scenarios where the compiler is able to devirtualize.

felipecrv commented on cpp/src/arrow/array/array_nested.h:
```diff
@@ -181,7 +199,14 @@ class ARROW_EXPORT LargeListArray : public BaseListArray<LargeListType> {
+  /// offsets's null bitmap. But if a null_bitmap is provided, the offsets array
+  /// can't have nulls.
+  ///
+  /// If a null_bitmap is provided, the offsets array can't be a slice (i.e. an
+  /// array with offset() > 0).
```
It's an overlook. For some reason, my brain is having a really hard time with these 4 slightly similar docstrings.

felipecrv replied:
I amended the last commit and force-pushed.

felipecrv commented on cpp/src/arrow/array/array_nested.h:
```diff
@@ -216,6 +241,183 @@ class ARROW_EXPORT LargeListArray : public BaseListArray<LargeListType> {
+  /// can't have nulls.
+  ///
+  /// If a null_bitmap is provided, the offsets array can't be a slice (i.e. an
+  /// array with offset() > 0).
```
No. I had the right comment in `LargeListView::FromArrays`. Copying it here and force pushing.

felipecrv commented on cpp/src/arrow/array/array_nested.h:
```diff
@@ -216,6 +241,183 @@ class ARROW_EXPORT LargeListArray : public BaseListArray<LargeListType> {
+  Result<std::shared_ptr<Array>> Flatten(
+      MemoryPool* memory_pool = default_memory_pool()) const;
+
+  /// \brief Return list-view offsets as an Int64Array
```
Pushing a commit fixing this.

felipecrv commented on cpp/src/arrow/array/concatenate.cc:
```diff
@@ -160,16 +166,69 @@ Status PutOffsets(const std::shared_ptr<Buffer>& src, Offset first_offset, Offse
+  for (size_t i = 0; i < buffers.size(); ++i) {
+    const auto displacement =
+        static_cast<offset_type>(num_child_values - value_ranges[i].offset);
+    PutListViewOffsets(/*src=*/*buffers[i], static_cast<offset_type>(displacement),
+                       /*dst=*/out_data + elements_length);
+    elements_length += buffers[i]->size() / sizeof(offset_type);
```
Oh, I think this is very personal, but I prefer code that doesn't mutate base buffer pointers and instead always re-derives the target pointer from the base pointer and an integer offset.

If you show me how you would leverage the returned pointer to simplify this loop I can totally change it though. I can't see it by myself.

felipecrv commented on cpp/src/arrow/array/concatenate_test.cc:
```diff
@@ -187,33 +189,89 @@ TEST_F(ConcatenateTest, FixedSizeListType) {
-    ArrayFromVector<Int64Type>(offsets_vector, &offsets);
-    ASSERT_OK_AND_ASSIGN(*out, LargeListArray::FromArrays(*offsets, *values));
+    auto offsets_vector = self.template Offsets<offset_type>(values_size, num_list_views);
+    offsets_vector.front() = 0;
```
Instead of improving these generators I propose we use `testing/random.h`. I created an issue for this now -> https://github.com/apache/arrow/issues/38656

But I would prefer tackling this problem in another PR. It's bigger than list-view: it would benefit the list concatenation tests as well.

felipecrv commented on cpp/src/arrow/array/util.cc:
```diff
@@ -383,6 +393,12 @@ class NullArrayFactory {
 
+    template <typename T>
+    enable_if_list_view<T, Status> Visit(const T&) {
+      buffer_length_ = length_ * sizeof(typename T::offset_type);
```
@bkietz rewrote this whole inner class in his PR didn't he? I hope I can merge this before his PR. :D

felipecrv replied:
Yeah. I also thought it should have been length 0 on the recursive call to children. I'm pushing a change with some `XXX` comments we should remove before merge.

felipecrv commented on cpp/src/arrow/array/util.cc:
```diff
@@ -853,6 +885,13 @@ class RepeatedArrayFactory {
   }
 
+  template <typename IntType>
+  Status CreateIntBuffer(IntType value, std::shared_ptr<Buffer>* out) {
```
Done. Pushing commit soon.

felipecrv commented on cpp/src/arrow/array/validate.cc:
```diff
@@ -23,7 +23,7 @@
 #include "arrow/type.h"
 #include "arrow/type_traits.h"
-#include "arrow/util/bit_block_counter.h"
+#include "arrow/util/bit_run_reader.h"
```
`validate.cc` didn't use any of the "counter" stuff. And after the spec simplifications that require null list-views to have valid dimensions, I don't even need the "reader" stuff anymore.

felipecrv commented on cpp/src/arrow/array/validate.cc:
```diff
@@ -797,57 +811,147 @@ struct ValidateArrayImpl {
+    const bool non_empty = data.length > 0;
+    if constexpr (is_list_view) {
+      if (!IsBufferValid(1)) {
+        // For length 0, an empty offsets buffer is accepted (ARROW-544).
```
Sure. Forgot to remove it when I changed the logic.

felipecrv commented on cpp/src/arrow/array/validate.cc:
```diff
@@ -797,57 +811,147 @@ struct ValidateArrayImpl {
   }
 
+ public:
+  template <typename TypeClass>
+  enable_if_list_view<TypeClass, Status> ValidateOffsetsAndSizes(const TypeClass& type,
+                                                                 int64_t offset_limit) {
+    return ValidateOffsetsAndMaybeSizes<TypeClass>(type, offset_limit);
+  }
+
+  template <typename TypeClass>
+  std::enable_if_t<is_var_length_list_type<TypeClass>::value ||
+                       is_base_binary_like(TypeClass::type_id),
+                   Status>
+  ValidateOffsets(const TypeClass& type, int64_t offset_limit) {
+    return ValidateOffsetsAndMaybeSizes<TypeClass>(type, offset_limit);
+  }
```
Two reasons:
 - Provide a simple public interface to the class
 - Constrain the types that can be passed as template parameters while all the internal private functions can assume the passed types are valid

felipecrv replied:
> But this class is just an implementation detail, it's not for outside use.

But in a 1000-line file like this I find it beneficial to be explicit about the functions that can only be called within the current class.

I will push a fixup with simplications.

felipecrv commented on cpp/src/arrow/c/bridge_test.cc:
```diff
@@ -1930,6 +2004,33 @@ TEST_F(TestSchemaImport, NestedList) {
+  FillListLike("+vl");
+  CheckImport(list_view(fixed_size_list(int8(), 3)));
+}
+
```
Oh yes. I've now completed the test set based on all occurrences of `List` in this file. It looks like this:

```
TEST_F(TestSchemaExport, ListView) {                TEST_F(TestSchemaExport, List) {
TEST_F(TestArrayExport, ListView) {                 TEST_F(TestArrayExport, List) {
TEST_F(TestArrayExport, ListViewSliced) {           TEST_F(TestArrayExport, ListSliced) {
TEST_F(TestDeviceArrayExport, ListView) {           TEST_F(TestDeviceArrayExport, List) {
TEST_F(TestDeviceArrayExport, ListViewSliced) {     TEST_F(TestDeviceArrayExport, ListSliced) {
TEST_F(TestSchemaImport, ListView) {                TEST_F(TestSchemaImport, List) {
TEST_F(TestSchemaImport, NestedListView) {          TEST_F(TestSchemaImport, NestedList) {
                                                    
TEST_F(TestArrayImport, ListView) {                 TEST_F(TestArrayImport, List) {
TEST_F(TestArrayImport, NestedListView) {           TEST_F(TestArrayImport, NestedList) {
TEST_F(TestArrayImport, ListViewWithOffset) {       TEST_F(TestArrayImport, ListWithOffset) {
TEST_F(TestArrayImport, ListViewNoError) {          TEST_F(TestArrayImport, ListError) {
TEST_F(TestSchemaRoundtrip, ListView) {             TEST_F(TestSchemaRoundtrip, List) {
TEST_F(TestArrayRoundtrip, ListView) {              TEST_F(TestArrayRoundtrip, List) {
```

Pushing commit soon.

felipecrv commented on cpp/src/arrow/ipc/writer.cc:
```diff
@@ -350,6 +350,67 @@ class RecordBatchSerializer {
+                          std::shared_ptr<Buffer>* out_value_sizes) {
+    const int64_t required_bytes = sizeof(offset_type) * array.length();
+    auto sizes = array.value_sizes();
+    if (sizes != nullptr && (array.offset() != 0 || sizes->size() > required_bytes)) {
```
Remember that in the beginning I was open to list-views having null buffers just like empty lists can have empty buffers? This is the reason. And I think keeping that check would be wise in case we revert that decision in the future.

This extra condition avoids an extra  `shared_ptr` allocation. Cheap, but something worth avoiding I guess. This was based on the code from `GetZeroBasedValueOffsets` introduced in 7f6e6ab2.

felipecrv commented on cpp/src/arrow/ipc/writer.cc:
```diff
@@ -442,6 +502,37 @@ class RecordBatchSerializer {
+    {
+      std::shared_ptr<Buffer> value_offsets;
+      RETURN_NOT_OK(
+          GetZeroBasedListViewOffsets<T>(array, &value_offsets, &min_offset, &max_end));
```
As above, I'm following the same logic in `GetZeroBasedValueOffsets`. This is how it is since the begining [1]. I don't know why.

[1] https://github.com/wesm/arrow/commit/df2220f350282925a454ed911eed6618e4d53969#diff-1b1d9dca9fdea7624e22f017b8762c4919edf57c2cf43c15d59b8a5e8e1b38a5R264

pitrou commented on cpp/src/arrow/array/array_nested.cc:
```diff
@@ -189,23 +260,126 @@ Result<std::shared_ptr<Array>> FlattenListArray(const ListArrayT& list_array,
+  }
+
+  std::vector<std::shared_ptr<Array>> slices;
+  {
```
This algorithm looks fine. It will be quite efficient if the list-view array is mostly contiguous and ascending, much less if the list view elements are all over the place :-)

Would you mind adding a comment about the perf expectations here?

(sidenote: perhaps we need a `Concatenate` API variant that takes a `vector<ArraySpan>`...)

felipecrv replied:
> Would you mind adding a comment about the perf expectations here?

Sure. Adding more text to the docstring.

> (sidenote: perhaps we need a Concatenate API variant that takes a vector<ArraySpan>...)

I once designed an API that had vectors of non-owning objects and it was a world of pain. We might come up with something better for Concatenate someday -- something that doesn't involve `std::vector` even.

pitrou commented on cpp/src/arrow/array/array_list_test.cc:
```diff
@@ -434,16 +622,18 @@ class TestListArray : public ::testing::Test {
     std::vector<int16_t> values = {0, 1, 2, 3, 4, 5, 6};
-    std::vector<int> lengths = {3, 0, 4};
     std::vector<uint8_t> is_valid = {1, 0, 1};
     // Should be {0, 3, 3} given the is_valid array
```
Can we fix this comment?

felipecrv replied:
Is it OK if I delete it? (I pushed a fixup commit dong that).

pitrou replied:
+1

pitrou reviewed (commented):
Thanks again for the updates. Some answers and more suggestions below.

felipecrv commented on cpp/src/arrow/array/util.cc:
```diff
@@ -379,6 +389,15 @@ class NullArrayFactory {
       // values array may be empty, but there must be at least one offset of 0
       RETURN_NOT_OK(MaxOf(sizeof(typename T::offset_type) * (length_ + 1)));
+      // XXX(felipec): reviewers, is this correct?
+      RETURN_NOT_OK(MaxOf(GetBufferLength(type.value_type(), length_)));
```
Adding a commit manual (to ensure `clang-format`).

felipecrv commented on cpp/src/arrow/array/util.cc:
```diff
@@ -379,6 +389,15 @@ class NullArrayFactory {
+    enable_if_list_view<T, Status> Visit(const T& type) {
+      RETURN_NOT_OK(MaxOf(sizeof(typename T::offset_type) * length_));
+      // XXX(felipec): reviewers, is this correct?
       RETURN_NOT_OK(MaxOf(GetBufferLength(type.value_type(), length_)));
```
Fixup commit coming soon.

felipecrv commented on cpp/src/arrow/c/bridge_test.cc:
```diff
@@ -3631,6 +3824,13 @@ TEST_F(TestArrayRoundtrip, List) {
 
+TEST_F(TestArrayRoundtrip, ListView) {
+  TestWithJSON(list_view(int32()), "[]");
+  TestWithJSON(list_view(int32()), "[[4, 5], [6, null], null]");
```
Isn't that an overkill given that round-tripping is all about passing the buffers around correctly without any consideration about their contents?

pitrou replied:
Fair enough :-)

felipecrv replied:
Alright, I just added a new test example in a fixup commit.

bkietz commented on cpp/src/arrow/array/util.cc:
```diff
@@ -379,6 +389,15 @@ class NullArrayFactory {
+    template <typename T>
+    enable_if_list_view<T, Status> Visit(const T& type) {
+      RETURN_NOT_OK(MaxOf(sizeof(typename T::offset_type) * length_));
+      // XXX(felipec): reviewers, is this correct?
```
Looks right to me

felipecrv replied:
Removed the XXX comments and pushed again.

felipecrv commented on cpp/src/arrow/array/concatenate_test.cc:
```diff
@@ -69,33 +99,117 @@ class ConcatenateTest : public ::testing::Test {
+    return rag_.ArrayOf(std::move(type), size, null_probability);
+  }
+
+  // TODO(GH-38656): Use the random array generators from testing/random.h here
```
@pitrou I isolated all the random-generation code in this class and removed the complicated `List[View]ConcatenationChecker` templates.

felipecrv commented on cpp/src/arrow/array/concatenate_test.cc:
```diff
@@ -69,33 +99,117 @@ class ConcatenateTest : public ::testing::Test {
+    auto values = PrimitiveArray<Int8Type>(values_size, null_probability);
+    auto offsets_vector = Offsets<offset_type>(values_size, length);
+    // Ensure first and last offsets encompass the whole values array
+    offsets_vector.front() = 0;
```
@pitrou I also don't do this for list-views anymore.

pitrou replied:
Thanks!

felipecrv commented:
@pitrou

pitrou commented on cpp/src/arrow/array/concatenate.cc:
```diff
@@ -160,16 +168,144 @@ Status PutOffsets(const std::shared_ptr<Buffer>& src, Offset first_offset, Offse
+    internal::BitBlockCount block = bit_counter.NextBlock();
+    if (block.AllSet()) {
+      for (int64_t i = 0; i < block.length; ++i, ++position) {
+        if (sizes[position] > 0) {
+          // NOTE: Concatenate can be called during IPC reads to append delta
+          // dictionaries. Avoid UB on non-validated input by doing the addition in the
+          // unsigned domain. (the result can later be validated using
+          // Array::ValidateFull)
+          const auto displaced_offset = SafeSignedAdd(offsets[position], displacement);
+          // displaced_offset>=0 is guaranteed by RangeOfValuesUsed returning the
+          // smallest offset of valid and non-empty list-views.
+          DCHECK_GE(displaced_offset, 0);
+          dst[position] = displaced_offset;
+        } else {
+          // Do nothing to leave dst[position] as 0.
+        }
```
I might be misreading, but is it just the same as `visit_not_null(i)`?

felipecrv replied:
You're right. I extracted the function from below when I noticed the dup, but forgot to do the reverse-inlining above.

Pushing soon.

pitrou commented on cpp/src/arrow/util/list_util.h:
```diff
@@ -0,0 +1,55 @@
+
+/// \brief Calculate the sum of the sizes of all valid lists or list-views
+///
+/// This is usally the same as the length of the RangeOfValuesUsed() range, but
```
```suggestion
/// This is usually the same as the length of the RangeOfValuesUsed() range, but
```

pitrou reviewed (approved):
Sorry for finding two more nits. Feel free to ping when done!

pitrou commented:
@felipecrv We'll want to update https://github.com/apache/arrow/blob/main/docs/source/status.rst in a followup PR.

felipecrv commented:
> @felipecrv We'll want to update https://github.com/apache/arrow/blob/main/docs/source/status.rst in a followup PR.

I will be extremely glad to send that PR.

mapleFU commented:
bravo 🍺!

mapleFU commented:
I've create an issue about parquet. https://github.com/apache/arrow/issues/38849
